src/Shared/OpportunisticIntern.cs

```diff
@@ -21,156 +17,83 @@ namespace Microsoft.Build
-            /// </summary>
-            char this[int index] { get; }
+        private static Dictionary<string, int> s_missedHardcodedStrings;
 
-            /// <summary>
-            /// Convert target to string. Presumed to be slow (and will be called just once).
-            /// </summary>
-            string ExpensiveConvertToString();
-
-            /// <summary>
-            /// Compare target to string. Assumes string is of equal or smaller length than target.
-            /// </summary>
-            bool StartsWithStringByOrdinalComparison(string other);
-
-            /// <summary>
-            /// Reference compare target to string. If target is non-string this should return false.
-            /// </summary>
-            bool ReferenceEquals(string other);
-        }
         #endregion
 
-        /// <summary>
-        /// Assign an int from an environment variable. If its not present, use the default.
-        /// </summary>
-        internal static int AssignViaEnvironment(string env, int @default)
-        {
-            string threshold = Environment.GetEnvironmentVariable(env);
-            if (!string.IsNullOrEmpty(threshold))
-            {
-                if (int.TryParse(threshold, out int result))
-                {
-                    return result;
-                }
-            }
-
-            return @default;
-        }
-
         /// <summary>
         /// Turn on statistics gathering.
         /// </summary>
         internal static void EnableStatisticsGathering()
         {
-            // Statistics include several 'what if' scenarios such as doubling the size of the MRU lists.
-            s_si = new BucketedPrioritizedStringList(/*gatherStatistics*/ true, s_smallMruSize, s_largeMruSize, s_hugeMruSize, s_smallMruThreshold, s_largeMruThreshold, s_hugeMruThreshold, s_ginormousThreshold, s_useSimpleConcurrency);
-            s_whatIfInfinite = new BucketedPrioritizedStringList(/*gatherStatistics*/ true, int.MaxValue, int.MaxValue, int.MaxValue, s_smallMruThreshold, s_largeMruThreshold, s_hugeMruThreshold, s_ginormousThreshold, s_useSimpleConcurrency);
-            s_whatIfDoubled = new BucketedPrioritizedStringList(/*gatherStatistics*/ true, s_smallMruSize * 2, s_largeMruSize * 2, s_hugeMruSize * 2, s_smallMruThreshold, s_largeMruThreshold, s_hugeMruThreshold, s_ginormousThreshold, s_useSimpleConcurrency);
-            s_whatIfHalved = new BucketedPrioritizedStringList(/*gatherStatistics*/ true, s_smallMruSize / 2, s_largeMruSize / 2, s_hugeMruSize / 2, s_smallMruThreshold, s_largeMruThreshold, s_hugeMruThreshold, s_ginormousThreshold, s_useSimpleConcurrency);
-            s_whatIfZero = new BucketedPrioritizedStringList(/*gatherStatistics*/ true, 0, 0, 0, s_smallMruThreshold, s_largeMruThreshold, s_hugeMruThreshold, s_ginormousThreshold, s_useSimpleConcurrency);
+            s_missedHardcodedStrings = new Dictionary<string, int>();
+            s_gatherStatistics = true;
```

Comment:
Do you guys ever need to gather stats on customer machines? Assuming not -- and this is just for tuning the interning, which you've made need much less tuning than before -- I suggest to put all the stats gathering code behind #if DEBUG and stop shipping it to customers.

Reply:
It's a valid point but I think I'll keep it in for now.

In the last couple of commits I brought the old implementation back as we agreed that it would be safer to provide an "escape hatch" in case the new one doesn't perform well, at least for a few releases. I've actually tried to put it all under DEBUG and it turned out that it would be either messy or require even more refactoring of code which we want to kill soon. Also, one of the numbers included in the statistics is "Time Spent Interning" which is only useful in release builds (although it's debatable whether it's useful at all given its implementation).